theano/tensor/nnet/neighbours.py

```diff
@@ -111,6 +111,43 @@ def grad(self, inp, grads):
                 return [neibs2images(gz, neib_shape, x.shape, mode=self.mode),
                         grad_undefined(self, 1, neib_shape),
                         grad_undefined(self, 2, neib_step)]
+
+        if self.mode in ['valid']:
```

Comment:
I just saw, the scan iterate on the number of pixel in the windows, not for each elements. That is good, it do a good chunk of work each time.

Be sure to test overlapping and case where neib_step > neib_shape. I think it is ok, but it is good to test that case too.